wp-includes/service-workers.php

```diff
@@ -0,0 +1,112 @@
+<?php
+/**
+ * Service Worker functions.
+ *
+ * @since ?
+ *
+ * @package PWA
+ */
+
+/**
+ * Initialize $wp_service_workers if it has not been set.
+ *
+ * @global WP_Service_Workers $wp_service_workers
+ * @return WP_Service_Workers WP_Service_Workers instance.
+ */
+function wp_service_workers() {
+	global $wp_service_workers;
+	if ( ! ( $wp_service_workers instanceof WP_Service_Workers ) ) {
+		$wp_service_workers = new WP_Service_Workers();
+	}
+	return $wp_service_workers;
+}
+
+/**
+ * Register a new service worker.
+ *
+ * @since ?
+ *
+ * @param string $handle   Name of the service worker. Should be unique.
+ * @param string $callback Callback method or relative path of the service worker.
+ * @param array  $deps     Optional. An array of registered script handles this depends on. Default empty array.
+ * @param array  $scopes   Optional Scopes of the service worker.
+ * @return bool Whether the script has been registered. True on success, false on failure.
+ */
+function wp_register_service_worker( $handle, $callback, $deps = array(), $scopes = array() ) {
+	$service_workers = wp_service_workers();
+
+	// If the path is not correct.
+	if ( ! is_callable( $callback ) && is_wp_error( $service_workers->get_validated_file_path( $callback ) ) ) {
+		_doing_it_wrong(
```

Comment:
This condition would probably be better moved to `WP_Service_Workers::register()`, no?